tpm2/platform.go

```diff
@@ -111,11 +144,115 @@ func (h *platformKeyDataHandler) RecoverKeys(data *secboot.PlatformKeyData) (sec
 }
 
 func (h *platformKeyDataHandler) RecoverKeysWithAuthKey(data *secboot.PlatformKeyData, key []byte) (secboot.KeyPayload, error) {
-	return nil, fmt.Errorf("passphrase authentication is not supported for the %s platform", platformName)
+	return h.recoverKeysCommon(data, key)
 }
 
 func (h *platformKeyDataHandler) ChangeAuthKey(handle, old, new []byte) ([]byte, error) {
-	return nil, fmt.Errorf("passphrase authentication is not supported for the %s platform", platformName)
+	tpm, err := ConnectToTPM()
+	switch {
+	case err == ErrNoTPM2Device:
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorUnavailable,
+			Err:  err}
+	case err != nil:
+		return nil, xerrors.Errorf("cannot connect to TPM: %w", err)
+	}
+	defer tpm.Close()
+
+	var k *SealedKeyData
+	if err := json.Unmarshal(handle, &k); err != nil {
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  err}
+	}
+	if k.data.Version() < 3 {
+		// All KeyData objects created for this platform are at least v3, so this
+		// should never fail. Test it though to avoid a panic later on if the version
+		// has been manipulated.
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  fmt.Errorf("invalid key data version: %d", k.data.Version())}
+	}
+
+	// Validate the initial key data
+	_, err = k.validateData(tpm.TPMContext, tpm.HmacSession())
+	switch {
+	case isKeyDataError(err):
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  err}
+	case err != nil:
+		return nil, xerrors.Errorf("cannot validate key data: %w", err)
+	}
+
+	srk, err := tpm.CreateResourceContextFromTPM(tcg.SRKHandle)
+	switch {
+	case tpm2.IsResourceUnavailableError(err, tcg.SRKHandle):
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorUninitialized,
+			Err:  ErrTPMProvisioning}
+	case err != nil:
+		return nil, xerrors.Errorf("cannot create context for SRK: %w", err)
+	}
+
+	keyObject, err := k.load(tpm.TPMContext, srk, tpm.HmacSession())
+	switch {
+	case isLoadInvalidParamError(err) || isImportInvalidParamError(err):
+		// The supplied key data is invalid or is not protected by the supplied SRK.
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorInvalidData,
+			Err:  xerrors.Errorf("cannot load sealed key object into TPM: %w", err)}
+	case isLoadInvalidParentError(err) || isImportInvalidParentError(err):
+		// The supplied SRK is not a valid storage parent.
+		return nil, &secboot.PlatformHandlerError{
+			Type: secboot.PlatformHandlerErrorUninitialized,
+			Err:  ErrTPMProvisioning}
+	case err != nil:
+		// This is an unexpected error
+		return nil, xerrors.Errorf("cannot load sealed key object into TPM: %w", err)
+	}
+	defer tpm.FlushContext(keyObject)
+
+	if len(old) > 0 {
+		v, err := deriveAuthValue(old, keyObject.Name().Algorithm().Size())
```

Comment:
but indeed here the code does something else, does the derivation turns an empty passphrase into an empty key? otherwise this doesn't seem quite right

Reply:
`old` in this context is not a passphrase - it's a passphrase derived key. It always has a non-zero size when a passphrase is set, regardless of the length of the passphrase (which could be zero).